source/extensions/filters/http/jwt_authn/authenticator.cc

```diff
@@ -147,6 +188,111 @@ void AuthenticatorImpl::verify(Http::HeaderMap& headers, Authenticator::Callback
   fetchRemoteJwks();
 }
 
+bool AuthenticatorImpl::setupPendingTokens() {
+  ASSERT(!grp_requirement_.empty());
+  if (!tokens_.empty()) {
+    return true;
+  }
+
+  std::vector<JwtRequirement>& req_list = grp_requirement_.back().second;
+  ASSERT(!req_list.empty());
+  while (containsList(req_list.back())) {
+    const auto& req = req_list.back();
+    unpackList(req);
+    req_list.pop_back();
+
+    req_list = grp_requirement_.back().second;
+    if (req_list.empty()) { // no more pending on current row
+      doneWithStatus(Status::Ok);
+      return false;
+    }
+  }
+
+  ASSERT(!req_list.empty());
+  const JwtRequirement req = req_list.back();
+  req_list.pop_back();
+
+  ok_to_bypass_ = req.has_allow_missing_or_failed() && req.allow_missing_or_failed().value();
+
+  TokenPredicate token_predicate;
+  Status status = Status::JwtMissed;
+  const auto& map = config_->getProtoConfig().providers();
+  setupAudienceOverride(req);
+  if (req.requires_type_case() == JwtRequirement::RequiresTypeCase::kProviderName ||
```

Comment:
In general,  I don't like this implementation.   It put all logic inside this class.  The code is fairly complicated.  

Please take a look rbac implementation in Envoy. https://github.com/envoyproxy/envoy/tree/master/source/extensions/filters/common/rbac

Please try to see if we can implement the matcher like that.